### ISSUE TITLE
Android notification stop button not working as expected

### ISSUE THREAD

**Issue by efraimbart** [ryanheise/just_audio#545]
**Which API doesn't behave as documented, and how does it misbehave?**
When using the notification stop button on Android the media stops appropriately, but then the notification returns to the playing state and, consequently, attempting to restart the media from the notification does not work.

**Minimal reproduction project**
Provide a link here using one of two options:
The example

**To Reproduce (i.e. user steps, not code)**
1. Launch the app
2. Click Play
3. Open the notification shade
4. Click Stop on the notification
5. Notice the notification goes to the stopped state for a second before reverting to the paused/playing state
6. Click Play on the notification
7. Notice that it has no effect

**Error messages**

**Expected behavior**
Like in the audio_audio service example, pressing stop should put the notification in the stopped state and pressing play should revive it.

**Screenshots**

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**
 - Device: Samsung Galaxy Note20 Ultra 5G
 - OS: Android 11 (One UI 3.1)

**Flutter SDK version**
```
[√] Flutter (Channel stable, 2.5.2, on Microsoft Windows [Version 10.0.22000.194], locale en-US)
[√] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
[√] Chrome - develop for the web
[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.11.2)
[√] Android Studio (version 2020.3)
[√] VS Code (version 1.60.0)
[√] Connected device (5 available)
```

**Additional context**

**Comment by ryanheise** [comment]
Can you try the different examples in the audio_service repo to see if any of them have the behaviour you desire? If so, let me know which one and I can copy that behaviour into just_audio_background.

**Comment by efraimbart** [comment]
Thanks! `audio_service/example/lib/main.dart` is the one I've been comparing against, and it seems to be the perfect behavior.

**Comment by ryanheise** [comment]
Thanks, I will work on a fix.

**Comment by efraimbart** [comment]
Thanks, will do!

**Comment by efraimbart** [comment]
Yes, though I'd still like to double check that. Before we go there though, are you on Android 11? Is there any chance you have the [media resumption](https://developer.android.com/guide/topics/media/media-controls#supporting_playback_resumption) feature disabled on your phone?

To be clear, I'm pretty sure this is an Android feature that holds on to the media session and allows resumption if playback is stopped from the app or by the system shutting it down. It's also what I was referring to in [this](https://github.com/ryanheise/just_audio/issues/545#issuecomment-941104570) comment.

**Comment by ryanheise** [comment]
Hi @efraimbart , the `dart.dart` example should demonstrate how to dismiss the notification while the `example_multiple_handler.dart` example should demonstrate the media session resumption feature, both under Android 11. When Android 11 first came out, the first example didn't work correctly, and that was due to a bug in Android itself that was later fixed. When the bug existed, they were apparently so intent on forcing media notifications to hang around that they actually made it impossible for an app to dismiss their own notification. That is no longer the case, at least on my Google Pixel 3a. Now of course different vendors may have their own versions of Android, so it's possible Samsung still has an issue.

But please do test the latest git commit in all cases to confirm that this is happening with the latest code.

> Thanks! `audio_service/example/lib/main.dart` is the one I've been comparing against, and it seems to be the perfect behavior.

I am surprised that this would be the perfect behaviour for you now that I know you actually want the media resumption feature. So one thing I'd appreciate if you could test is whether the `main.dart` example really can resume playback after stop (because this example is specifically programmed not to do that!)

**Comment by efraimbart** [comment]
Hey @ryanheise, I've since tested the latest version of audio_service (on `master`, commit 50bf147df95dea649d4d47b622d4d61033e4a75c, main.dart - using just `flutter run`) as well as the just_audio fix for background stop (`fix/background_stop`, commit 7fd18d52906716e138e2e3085290f37b7de3a2df), both on a clean image of Android 11 on a Pixel 5 emulator. Beforehand I wiped my pub.cache folder and reset the repositories just in case. 

With both of them, stopping the music didn't clear the notification - as seen in the previously attached video, but with audio_service I was able to restart the media from the notification after stopping it, while I was unable to in just_audio.

Thanks for taking the time to look into this!

**Comment by nt4f04uNd** [comment]
Ah you answered this, sorry I overlooked, the recents

**Comment by ryanheise** [comment]
@nt4f04uNd are you able to check whether `notify()` is actually being called after `cancel()` in your case? Hopefully that's it because that can be fixed.

**Comment by ryanheise** [comment]
Also, I'd be interested to see if any apps on the Play Store exhibit the same behaviour.

For example, when I try [Pi Music Player](https://play.google.com/store/apps/details?id=com.Project100Pi.themusicplayer), it provides an `x` action in the notification which (for me) cancels the notification successfully. Can you try it and let me know if you're noticing the same failure on your device?

On the other hand, if it successfully cancels the notification on Pi Music Player but not in audio_service's `lib/main.dart`, then there's something in audio_service that could be improved.

**Comment by ryanheise** [comment]
Possibly. In my local copy I have it guarded like this:

```java
    private void deactivateMediaSession() {
        if (mediaSession.isActive()) {
            mediaSession.setActive(false);
        }
        // Force cancellation of the notification
        // XXX: Maybe this should be done only for stop() and not for onDestroy(), to help preserve the notification layout?
        if (notificationCreated) {
            getNotificationManager().cancel(NOTIFICATION_ID);
            notificationCreated = false;
        }
    }
```

**Comment by ryanheise** [comment]
> Should we break this up into two issues

Yes, there are definitely two issue here. I believe I've fixed things on the just_audio_background side, but will leave this open until the audio_service issue is resolved, too.

> Just tried swapping deactivateMediaSession() and stopSelf() and it didn't seem to do anything.
Perhaps attempt with a clean emulator and see if it's still different for you?

I wouldn't expect that to do anything, but who knows? It could be a fragile behaviour that we have to work around. Calling `cancel` twice isn't good, hence my guard above. But I'd like to see whether this is actually possible to fix by testing on other apps. like Pi Music Player.

**Comment by nt4f04uNd** [comment]
> @nt4f04uNd are you able to check whether notify() is actually being called after cancel() in your case? Hopefully that's it because that can be fixed.

It's not the case, I get

```
stopForeground(false)
CANCEL
CANCEL
CANCEL
stopForeground(false)
```

The stumbled upon a media resumption article https://developer.android.com/guide/topics/media/media-controls.

Now what actually worked is adding this to `onGetRoot`

```java
if (isRecentRequest) {
    return null;
}
```

The `clientPackageName` for this is `com.android.systemui`

Before discovering this I also tried setting an empty `title` and `album` which left a dead notification when the service is stopped, which then cannot be updated even after the app restart LOL

**Comment by ryanheise** [comment]
Oh right, the actual recents are returned by on load children. But there is still a way to get an instant results from on get root and that's to prepare the response in a config parameter.

**Comment by nessblo** [comment]
Hi!
I have this issue as well with just_audio 0.9.17 and audio service0.18.1. The notification does not disappear

**Comment by nessblo** [comment]
Hello, 
1- how can i use this fix,
2- In exemple_playilist I saw  ""in this example, the service stops when reaching the end"", how can i change this

Excuse me I am still a beginner and I do not understand some things well.
Thanks!

**Comment by nessblo** [comment]
> Hello, 1- how can i use this fix, 2- In exemple_playilist I saw ""in this example, the service stops when reaching the end"", how can i change this
> 
> Excuse me I am still a beginner and I do not understand some things well. Thanks!

And how to update queue and playlist please.

**Comment by ryanheise** [comment]
I'm taking another look at this.

> With both of them, stopping the music didn't clear the notification - as seen in the previously attached video,

But your video above is not of `main.dart`. I would be interested to see a video exhibiting this behaviour specifically of `main.dart`.

**Comment by efraimbart** [comment]
Just to clarify before I dive back into this, if I recall correctly this was all done in main.dart, what are you seeing that suggests otherwise?

**Comment by ryanheise** [comment]
I've just merged and published what was in the `fix/background_stop` branch since it is at least an improvement over the previous beta and it worked perfectly for me. i.e. the notification was removed on Android after clicking stop. If you can share a video of `main.dart` showing it not working, I'll try to investigate any remaining issue.

**Comment by ryanheise** [comment]
(by which I mean `audio_service/example/lib/main.dart`)

**Comment by ryanheise** [comment]
May I know whether there is still an issue?

**Comment by efraimbart** [comment]
I haven't had the chance to jump back into this for a while now, so feel free to close and we can reopen to discuss again if it comes back up.

**Comment by ryanheise** [comment]
Sounds like a plan :+1:
